scintillometry/tests/test_folding.py

```diff
@@ -0,0 +1,90 @@
+# Licensed under the GPLv3 - see LICENSE
+"""Full-package tests of pulse folding."""
+
+import pytest
+import numpy as np
+import astropy.units as u
+from astropy.time import Time
+
+from ..base import Task
+from ..generators import EmptyStreamGenerator
+from ..pulse_folding import TimeFold
+from ..functions import Square
+
+
+class TestFoldingBase:
+    def setup(self):
+        self.start_time = Time('2010-11-12T13:14:15')
+        self.sample_rate = 10. * u.kHz
+        self.shape = (6000, 2, 4)
+        self.nh = EmptyStreamGenerator(shape=self.shape,
+                                       start_time=self.start_time,
+                                       sample_rate=self.sample_rate,
+                                       samples_per_frame=200, dtype=np.float)
+        self.sh = Task(self.nh, self.pulse_simulate, samples_per_frame=1)
+        self.period_bin = 125
+        self.phase_bin = 50
+
+    def phase(self, t):
+        dt = (t - t[0]).to(u.s)
+        F0 = 1.0 / (self.period_bin / self.sh.sample_rate)
+        return F0 * dt
+
+    def pulse_simulate(self, fh, data):
+        idx = fh.tell()
+        data [:]= 0
+        data += (10 if idx % self.period_bin ==0 else 0.125)
+        return data
+
+
+class TestTimeFolding(TestFoldingBase):
+    def test_input_data(self):
+        indata = self.sh.read(1000)
+        pulses = np.where(indata[:, 0, 0] == 10)[0]
+        # Check if the input data is set up right.
+        assert (len(pulses) == 8), "Pulses are not simulated right."
+
+    def test_under_period(self):
+        # Test when folding time smaller than the pulse period.
+        self.fold_time = 11 * u.ms
+        eff_phase_bin = (11* u.ms / (self.period_bin / self.sh.sample_rate) * \
+                         self.phase_bin).to(u.Unit(1))
+        self.fh = TimeFold(self.sh, self.phase_bin, self.fold_time, \
+                           self.phase, samples_per_frame=20)
+        self.fh.seek(0)
+        fr = self.fh.read(3)
+
+        for ii in range(self.fh.counts.shape[0]):
```

Comment:
Would do `for ii, count in enumerate(self.fh.counts)`.

Obviously, have a bit of a problem here in using API that we hope not to maintain... One solution would be to make the pulse profile have a base (i.e., 10 on pulse, 1 out of pulse), and then check that there are zeros in the result.

Reply:
The simulated pulse will occur every 125 sample bins. Is this solves the concern?